File: server-node/src/application/use-cases/conversation/get-user-conversations.ts
import { Injectable, Logger } from '@nestjs/common';

import { User } from '@application/entities/user';
import { Message } from '@application/entities/message';

import {
  IGetConversationsByUserIdResponse,
  IGetParticipantsByConversationIdResponse,
  ParticipantRepository,
} from '@application/repositories/participant-repository';
import { MessageRepository } from '@application/repositories/message-repository';
import { MessageStatusRepository } from '@application/repositories/message-status-repository';

import { Either, left, right } from '@helpers/Either';
import { IMessageViewModel } from '@infra/http/view-model/message-view-model';

interface Request {
  userId: User['_id'];
}

interface Response extends IGetConversationsByUserIdResponse {
  participants: IGetParticipantsByConversationIdResponse[];
  lastMessages: IMessageViewModel[];
  unreadMessagesCount: number;
}

@Injectable()
export class GetUserConversationsUseCase {
  constructor(
    private readonly participantRepository: ParticipantRepository,
    private readonly messageRepository: MessageRepository,
    private readonly messageStatusRepository: MessageStatusRepository,
  ) {}

  async execute({ userId }: Request): Promise<Either<Error, Response[]>> {
    const conversationsThatUserParticipate = await this.participantRepository.getConversationsByUserId(userId);

    if (conversationsThatUserParticipate.isLeft()) {
      return left(new Error('Error on try to get conversations from database'));
    }

    const conversations = await Promise.all(
      conversationsThatUserParticipate.value.map(async conversation => {
        const [participants, lastMessages, unreadMessagesCount] = await Promise.all([
          await this.participantRepository.getParticipantsByConversationId(conversation.id),
          await this.messageRepository.getMessagesFromConversation(conversation.id, 3),

          conversation.isGroup
            ? await this.messageStatusRepository.getUnreadMessagesStatusCountByUserId(userId, conversation.id) // Group Conversation
            : await this.messageRepository.getUnreadConversationMessagesCount(userId, conversation.id), // Private Conversation
        ]);

        if (participants.isLeft()) Logger.error(`Conversation: ${conversation.id}`, participants.value);
        if (lastMessages.isLeft()) Logger.error(`Conversation: ${conversation.id}`, lastMessages.value);

        return {
          ...conversation,
          unreadMessagesCount: unreadMessagesCount?.isRight() ? unreadMessagesCount.value : 0,
          participants: participants.isRight() ? participants.value : [],
          lastMessages: lastMessages.isRight() ? lastMessages.value : [],
        };
      }),
    );

    return right(conversations);
  }
}

File: Proxymity-Chat-Service-Go/internal/application/services/conversation_service/get_user_conversations.go
package services

import (
	"context"
	"proxymity/chat-server/internal/domain/entities"
	"proxymity/chat-server/internal/infra/database/repositories"
	"sort"
	"time"

	"github.com/google/uuid"
)

type GetUserConversationsRequest struct{}
type GetUserConversationsResponse struct {
	Conversation        repositories.GetConversationsByUserId          `json:"conversation"`
	UnreadMessagesCount int                                            `json:"unreadMessagesCount"`
	Participants        []repositories.GetParticipantsByConversationId `json:"participants"`
	LastMessages        []*entities.Message                            `json:"lastMessages"`
}

func (cs *ConversationService) GetUserConversations(
	ctx context.Context,
	userId uuid.UUID,
) ([]GetUserConversationsResponse, error) {
	conversationsThatUserParticipate, err := cs.participantRepository.GetConversationsByUserId(ctx, userId)

	if err != nil {
		return nil, err
	}

	conversations := make([]GetUserConversationsResponse, len(*conversationsThatUserParticipate))

	for i, conversation := range *conversationsThatUserParticipate {
		participants, err := cs.participantRepository.GetParticipantsByConversationId(ctx, conversation.Id)

		if err != nil {
			return nil, err
		}

		lastMessages, err := cs.messageRepository.GetMessagesFromConversation(ctx, conversation.Id, 1)

		if err != nil {
			return nil, err
		}

		var unreadMessagesCount int

		if conversation.GroupId != nil {
			cs.messageRepository.GetUnreadConversationMessagesCount(ctx, conversation.Id, userId)
		} else {
			cs.messageRepository.GetUnreadConversationMessagesCount(ctx, conversation.Id, userId)
		}

		conversations[i] = GetUserConversationsResponse{
			Conversation:        conversation,
			UnreadMessagesCount: unreadMessagesCount,
			Participants:        *participants,
			LastMessages:        lastMessages,
		}

		sort.Slice(conversations, func(i, j int) bool {
			var currDate, prevDate time.Time

			if len(conversations[i].LastMessages) > 0 {
				currDate = conversations[i].LastMessages[0].WrittenAt
			} else {
				currDate = time.Time{}
			}

			if len(conversations[j].LastMessages) > 0 {
				prevDate = conversations[j].LastMessages[0].WrittenAt
			} else {
				prevDate = time.Time{}
			}

			return prevDate.Before(currDate)
		})
	}

	return conversations, nil
}
